ath79: fix irq cascade driver and add wmac for ar934x

981213 opened the pull request:
Add the missing enable and disable function.
Remove dummy mask and unmask function and use the one provided by irq_dummy_chip.
Allow interrupt status register being defined from dts and add ddr_wb_flush for ar934x/qca953x.
Tested on Qihoo C301 (AR9344 + QCA9882), both PCIE and WMAC works fine.

981213 commented:
PS: If this got applied, the intc3 node of qca956x.dtsi from #1043 should also be changed.

juppin commented:
Firstly, thanks for digging into this problem.

Secondly, you forgot to remove the wb flush of interrupt 2 of cpuintc node in ar934x.dtsi.
The wb flush will happen currently two times...

Thirdly, wouldn´t it a better idea to implement the dt definition for "qcom,ath79-intc" the same way as for  "qca,ar7100-cpu-intc".
I mean with a phandle to ddr registers and the wb flush number/offset...?

And lastly, as we dont´t need "qcom,ath79-intc" for ar9341 because cpu intc 2 is only used for wmac, it would be better to use the cascade driver only for devices that have the shared int for pci and wmac. So only use the node definition in ar9344.dtsi and others with pci support.

981213 commented:
Thanks. I'll do those tomorrow.

hackpascal commented:
Tested on QCA9563 (Phicomm K2T).
Works fine.

981213 commented:
The ar9341.dtsi here isn't tested because I currently don't have one.

981213 commented:
BTW: I renamed all dt node names due to the following reason:
1 This interrupt design was introduced since ar934x
2 there is a node named '#qca,ddr-wb-channel-cells' which was defined in ddr_ctrl node and I simply wants the manufacturer look the same :D 
If anyone is against the rename about the manufacturer, I'll change it back.

juppin commented:
For me your work looks very good :-) Many thanks to you...
I´m totally with you about the naming.

> The ar9341.dtsi here isn't tested because I currently don't have one.

I have almost the same ar9341.dtsi in my local branch for my tp link wr841v8 and it works with no problems. 
Haven´t figured out how to get the switch working as in ar71xx, so thats the reason why i haven´t created a PR

981213 commented:
The following can get gmac1 and builtin switch work, However I'm unable to enable gmac0 which connected directly to phy0 or phy4 of the builtin switch and should be polling data from mdio1.
```
&mdio1 {
	status = "okay";
	phy4-mii-enable;

	phy4: ethernet-phy@4 {
		reg = <0>;
		phy-mode = "mii";
	};
};

&eth0 {
	status = "okay";
	phy-mode = "mii";

	phy-handle = <&phy4>;

	gmac-config {
		device = <&gmac>;
		switch-phy-swap = <1>;
	};
};

&eth1 {
	status = "okay";
	phy-mode = "gmii";

	fixed-link {
		speed = <1000>;
		full-duplex;
	};
};
```
dmesg here:
```
[    0.585894] ag71xx 19000000.eth: invalid MAC address, using random address
[    1.222662] ag71xx 19000000.eth: Could not connect to PHY device
[    1.228989] ag71xx 1a000000.eth: invalid MAC address, using random address
[    2.201930] libphy: mdio: probed
[    2.208751] mdio-bus.0: Found an AR934X built-in switch
[    2.255737] ag71xx 1a000000.eth: connected to PHY at fixed-0:00 [uid=00000000, driver=Generic PHY]
[    2.265486] eth0: Atheros AG71xx at 0xba000000, irq 5, mode:GMII
```

juppin commented:
> However I'm unable to enable gmac0 which connected directly to phy0 or phy4 of the builtin switch and should be polling data from mdio1.

Thats exacly the same part which i didn't figured out how to encode it in dt. I think with the current of drivers it is not possible to use mdio1 also for gmac0.
With a fixed link it does work but then eth0 is always up and didn't recognize a unplugged cable.

981213 commented:
@juppin Several problems I found about GMAC here:
GMAC1 should be registered before GMAC0 to get mdio1 and builtin-switch properly configured. (This can be done by swapping eth0 and eth1 node in ath79.dtsi however this will affect many devices currently supported by this target.)
phy-poll-mask for ag7240 is missing.
There is an missing register if we use ar9330-gmac for ar934x. ar933x doesn't have SW_ONLY_MODE (bit 6) and ar934x doesn't have SW_PHY_ADDR_SWAP(bit 8)

There are other problems I don't know :( With all those things I mentioned above fixed, the switch works weird. previously working switch is broken and plugging an ethernet cable to wan port seems triggered a switch reset. No idea about what's going on.

lucize commented:
can't de nodes from ath79.dtsi be redefined in other dtsi ?

lucize commented:
Yes, I used that, intc2 and 1, got assigned irq 11 in os but after wlan up, it hanged, i'll post the dts

lucize commented:
@981213 so adding to `qca9557.dtsi` for intc2 like on your `ar9344.dtsi` will not hang with the wifi anymore

```
		qca,ddr-wb-channel-interrupts = <0>, <1>;
		qca,ddr-wb-channels = <&ddr_ctrl 3>,
				<&ddr_ctrl 4>;
```
but I have panic on usb


```
			usb_phy0: usb-phy@18030000 {
				compatible ="qca,qca9550-usb-phy", "qca,ar7200-usb-phy";

				reset-names = "usb-phy", "usb-suspend-override";
				resets = <&rst 4>, <&rst 3>;

				#phy-cells = <0>;

				status = "disabled";
			};

```
and

```
		usb0: usb@1b000000 {
			compatible = "generic-ehci";
			reg = <0x1b000000 0x1fc>;

			interrupt-parent = <&intc3>;
			interrupts = <1>;
			resets = <&rst 5>;
			reset-names = "usb-host";

			has-transaction-translator;
			caps-offset = <0x100>;

			phy-names = "usb-phy0";
			phys = <&usb_phy0>;

			status = "disabled";
		};
```

```
[   62.448581] CPU: 0 PID: 0 Comm: swapper Not tainted 4.14.49 #0
[   62.454491] task: 804e19f0 task.stack: 804dc000
[   62.459080] $ 0   : 00000000 81110000 00000001 fffffffe
[   62.464386] $ 4   : c626c95c 00000024 804f0000 00000001
[   62.469692] $ 8   : 00000001 00019750 00000001 00000001
[   62.474999] $12   : ffffffff 0000002a 804f0000 2397f82a
[   62.480305] $16   : 83804b28 00000025 83803880 01000000
[   62.485611] $20   : ffffffff 00000003 00000000 83ffc000
[   62.490919] $24   : 804df020 8020d960
[   62.496225] $28   : 804dc000 83809f68 80060000 8020d47c
[   62.501532] Hi    : 00144a31
[   62.504444] Lo    : c74ad398
[   62.507369] epc   : 80061398 ath79_ddr_wb_flush+0x14/0x44
[   62.512862] ra    : 8020d47c ath79_intc_irq_handler+0x7c/0xe8
[   62.518683] Status: 1100fd02 KERNEL EXL
[   62.522662] Cause : 0080000c (ExcCode 03)
[   62.526726] BadVA : c626c95c
[   62.529640] PrId  : 00019750 (MIPS 74Kc)
[   62.533616] Modules linked in: ath9k ath9k_common pppoe ppp_async ath9k_hw ath pppox ppp_generic nf_conntrack_ipv6 mac80211 iptable_nat ipt_REJECT ipt_MASQUERADE cfg80211 xt_time xt_tcpudp xt_state xt_nat xt_multiport xt_mark xt_mac xt_limit xt_conntrack xt_comment xt_TCPMSS xt_REDIRECT xt_LOG xt_FLOWOFFLOAD slhc nf_reject_ipv4 nf_nat_redirect nf_nat_masquerade_ipv4 nf_conntrack_ipv4 nf_nat_ipv4 nf_nat nf_log_ipv4 nf_flow_table_hw nf_flow_table nf_defrag_ipv6 nf_defrag_ipv4 nf_conntrack_rtcache nf_conntrack iptable_mangle iptable_filter ip_tables crc_ccitt compat ledtrig_usbport ip6t_REJECT nf_reject_ipv6 nf_log_ipv6 nf_log_common ip6table_mangle ip6table_filter ip6_tables x_tables ehci_platform ehci_hcd gpio_button_hotplug usbcore nls_base usb_common
[   62.601204] Process swapper (pid: 0, threadinfo=804dc000, task=804e19f0, tls=00000000)
[   62.609225] Stack : 804f0b00 800b3b28 f0000000 ffffffff 00000003 804f09c0 f0000000 ffffffff
[   62.617707]         c0000000 800af0a0 80490000 00200102 00000080 804f09c0 f0000000 803f906c
[   62.626191]         c0000000 00000007 00000000 83ffc000 00000008 8020d3a0 e1966d86 e9b629b6
[   62.634674]         e9976d91 699425a3 29b6e196 00000000 804ddde0 804e0000 00000001 81110000
[   62.643158]         00000000 800658b8 804ddde0 61856d86 a9b56996 29866986
[   62.650049] Call Trace:
[   62.652526] [<80061398>] ath79_ddr_wb_flush+0x14/0x44
[   62.657653] [<8020d47c>] ath79_intc_irq_handler+0x7c/0xe8
[   62.663146] [<800af0a0>] generic_handle_irq+0x38/0x50
[   62.668282] [<803f906c>] do_IRQ+0x1c/0x2c
[   62.672347] [<8020d3a0>] plat_irq_dispatch+0xc0/0x120
[   62.677471] [<800658b8>] handle_int+0x138/0x144
[   62.682073] [<800673ac>] r4k_wait_irqoff+0x18/0x24
[   62.686936] [<800a85d0>] do_idle+0x78/0xe4
[   62.691089] [<800a8828>] cpu_startup_entry+0x18/0x20
[   62.696126] [<80500bc0>] start_kernel+0x454/0x474
[   62.700901] Code: 8c4220d4  00442021  24020001 <ac820000> 8c820000  30420001  1440fffd  24020001  ac820000
[   62.710803]
[   62.712361] ---[ end trace 0a39709bcabfe9a0 ]---
[   62.718566] Kernel panic - not syncing: Fatal exception in interrupt
[   62.725657] Rebooting in 3 seconds..

```

lucize commented:
adding the same for `intc3` solves the usb crash

```
		qca,ddr-wb-channel-interrupts = <0>, <1>;
		qca,ddr-wb-channels = <&ddr_ctrl 3>,
				<&ddr_ctrl 4>;
```

981213 commented:
There may be problems about the added ddr_wb_flush part in my patches which caused all those problems. I don't have much knowledge about dts and I just copied those code from irq-ath79-cpu.c. There isn't ddr-wb-channels defined for qca955x and qca956x in ar71xx and what I want to do is to ignore the call to ath79_ddr_wb_flush when ddr-wb-channel nodes for the interrupt are not present. Could someone help me check the code? I'm busy these days and I have to wait till this weekend to check this problem.

jneuhauser committed:
ath79: fix irq-ath79-intc driver and add support for other ath79 SoCs

Add the missing enable and disable function.
Remove dummy mask and unmask function and use the one provided by
irq_dummy_chip.
Allow interrupt status register being defined from dts.
Add ddr_wb_flush for ar934x/qca953x.
Rename controller name to qca,ar9340-intc because this design was
first introduced in AR934x.

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>

Signed-off-by: Johann Neuhauser <johann@it-neuhauser.de>

juppin commented on target/linux/ath79/patches-4.14/0007-irqchip-irq-ath79-intc-add-irq-cascade-driver-for-QC.patch:
```diff
@@ -95,27 +104,56 @@ Signed-off-by: John Crispin <john@phrozen.org>
++		intc->irq_wb_chan[i] = 0xffffffff;
++	}
++
++	cnt = of_count_phandle_with_args(
```
Here is the bug that @lucize mentioned...
If someone does not set "qca,ddr-wb-channelXXX" properties, this value is zero and 
`domain = irq_domain_add_linear(node, cnt, &ath79_irq_domain_ops, intc);`
the second param cnt (=hwirq_max) is also zero.

981213 replied:
Thanks for pointing out this!

juppin commented:
@lucize @hackpascal I´ve modified the intc driver as i commented above, modified the dtsi for qca9557 and qca956x and tested it with wr1043v3. But i could not confirm if pcie does work as expected... Could you try my branch on your devices? I´ve also added the second pcie rc node...
https://github.com/juppin/openwrt/tree/irq-ath79-intc

On ar9344_tl-wdr3600.dts and qca9558_tl-wr1043nd-v3.dts everything does now work with this driver!

lucize commented:
@juppin I was just trying to compile it so I can test it on v4 and uap-lite (wmal+pcie)

981213 commented:
@lucize The only pcie on qca956x is using interrupts splitted from IRQ3 and IRQ2 is only used by wmac. So we don't need intc2 for qca956x.

juppin commented:
@981213 At least for uap-lite, k2t and c7v2 this is true. But probably there is a device that also use the first pcie.

@lucize which address you changed on pcie1?
At least the second usb (usb1) will / should use the second reset node.

Will correct the missed forgotten wmac node and remove pcie1 node from all devices from my testing branch...
Should we not letft the pcie1 in qca956x.dtsi?

hackpascal commented:
@juppin qca956x has only one pcie (exposed pads), which is identically the second pcie on qca955x. So there is no possibility that a device uses the 'first pcie'.

juppin commented:
@hackpascal Thanks. 

I´ve updated my branch.

@981213 So if you fix the wb parse and add the changes for 9557.dtsi and 956x.dtsi this should be ready to get merged. Hopefully :-)

jneuhauser committed:
ath79: ar934x: use irq-ath79-intc driver and enable wmac

Signed-off-by: Johann Neuhauser <johann@it-neuhauser.de>

jneuhauser committed:
ath79: qca95xx: add new intc2, correct intc3 and add second pcie on qca9557

Signed-off-by: Johann Neuhauser <johann@it-neuhauser.de>

981213 commented:
@juppin I've just cherry-picked all your commits here. Thanks a lot!

juppin commented:
@blogic Any suggestions for improvements?

No one reported issues, except the issue if the properties for "qca,ddr-wb-channelXXX" are not set.
This was fixed by me and @981213 cherry-picked my commits into this pr.

The last pr state is currently tested on wr1043v3 (qca9558) and wdr3600 (ar9344) by me, on wr1043v4 (qca9563) and uap-lite (qca9563) by @lucize and the previous version without my fix for wb parse but with set "qca,ddr-wb-channelXXX" properties on Phicomm K2T (qca9563) by @hackpascal and on Qihoo C301 (ar9344) by @981213.

blogic commented:
merged, Thanks !

lucize commented:
@981213 

> GMAC1 should be registered before GMAC0 to get mdio1 and builtin-switch properly configured. 

how about create another ath79_swap.dtsi just for this soc and include that only that ?

blogic commented:
a simple swizzle should suffice by making gmac1 syscon only.

lucize commented:
tried syscon only on qca9533 and only one ag71xx device would be detected then

lucize commented:
@981213 PR https://github.com/openwrt/openwrt/pull/983 added support for QCA9533 but the dts was not updated to new driver, trying it on wr841n I get this:

```
[    0.000000] ------------[ cut here ]------------
[    0.000000] WARNING: CPU: 0 PID: 0 at mm/page_alloc.c:3877 0x800f1188
[    0.000000] Modules linked in:
[    0.000000] CPU: 0 PID: 0 Comm: swapper Not tainted 4.14.50 #0
[    0.000000] Stack : 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[    0.000000]         00000000 00000000 00000000 00000000 00000000 00000001 80477bd8 214d68a8
[    0.000000]         80477c70 00000000 00000000 000009f0 00000038 803f1a58 00000007 00000000
[    0.000000]         00000026 00000000 00000025 00000000 80477bb8 00000000 8042c4d0 800f1188
[    0.000000]         00000009 00000f25 8100360c 80477df0 00000001 80061450 81faffb0 fffffffe
[    0.000000]         ...
[    0.000000] Call Trace:
[    0.000000] [<803f1a58>] 0x803f1a58
[    0.000000] [<800f1188>] 0x800f1188
[    0.000000] [<80061450>] 0x80061450
[    0.000000] [<8006a31c>] 0x8006a31c
[    0.000000] [<8006a324>] 0x8006a324
[    0.000000] [<80080270>] 0x80080270
[    0.000000] [<800f1188>] 0x800f1188
[    0.000000] [<80080334>] 0x80080334
[    0.000000] [<800f1188>] 0x800f1188
[    0.000000] [<800b63a0>] 0x800b63a0
[    0.000000] [<800b65dc>] 0x800b65dc
[    0.000000] [<801098d8>] 0x801098d8
[    0.000000] [<800b5568>] 0x800b5568
[    0.000000] [<804a9840>] 0x804a9840
[    0.000000] [<800b5424>] 0x800b5424
[    0.000000] [<801054b8>] 0x801054b8
[    0.000000] [<800b3058>] 0x800b3058
[    0.000000] [<8020b198>] 0x8020b198
[    0.000000] [<800b337c>] 0x800b337c
[    0.000000] [<804a9948>] 0x804a9948
[    0.000000] [<804af49c>] 0x804af49c
[    0.000000] [<8049d34c>] 0x8049d34c
[    0.000000] [<80499978>] 0x80499978
[    0.000000] [<80499380>] 0x80499380
[    0.000000]
[    0.000000] ---[ end trace e993832516409e7f ]---
[    0.000000] ------------[ cut here ]------------
[    0.000000] WARNING: CPU: 0 PID: 0 at kernel/irq/irqdomain.c:141 0x800b53b4
[    0.000000] Modules linked in:
[    0.000000] CPU: 0 PID: 0 Comm: swapper Tainted: G        W       4.14.50 #0
[    0.000000] Stack : 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[    0.000000]         00000000 00000000 00000000 00000000 00000000 00000001 80477ca0 214d68a8
[    0.000000]         80477d38 00000000 00000000 00001118 00000038 803f1a58 00000007 00000000
[    0.000000]         0000004b 00000000 0000004a 00000000 80477c80 00000000 80428b14 800b53b4
[    0.000000]         00000009 0000008d 8100360c 80477df0 00000003 80061450 81faffb0 fffffffe
[    0.000000]         ...
[    0.000000] Call Trace:
[    0.000000] [<803f1a58>] 0x803f1a58
[    0.000000] [<800b53b4>] 0x800b53b4
[    0.000000] [<80061450>] 0x80061450
[    0.000000] [<8006a31c>] 0x8006a31c
[    0.000000] [<8006a324>] 0x8006a324
[    0.000000] [<80080270>] 0x80080270
[    0.000000] [<800b53b4>] 0x800b53b4
[    0.000000] [<80080334>] 0x80080334
[    0.000000] [<801098d8>] 0x801098d8
[    0.000000] [<800b53b4>] 0x800b53b4
[    0.000000] [<804a9840>] 0x804a9840
[    0.000000] [<800b5424>] 0x800b5424
[    0.000000] [<801054b8>] 0x801054b8
[    0.000000] [<800b3058>] 0x800b3058
[    0.000000] [<8020b198>] 0x8020b198
[    0.000000] [<800b337c>] 0x800b337c
[    0.000000] [<804a9948>] 0x804a9948
[    0.000000] [<804af49c>] 0x804af49c
[    0.000000] [<8049d34c>] 0x8049d34c
[    0.000000] [<80499978>] 0x80499978
[    0.000000] [<80499380>] 0x80499380
[    0.000000]
[    0.000000] ---[ end trace e993832516409e80 ]---
```

the wmac is starting if I exclude from `intc2` the `interrupt-parent = <&cpuintc>;` adingthe wmac to `intc2` would not work

changing only `compatible = "qca,ar9340-intc";` will panic with `[    0.000000] Kernel panic - not syncing: Missing address of interrupt status register`

the target is tiny so I think that I need to activate some more kernel debuging

981213 commented:
the intc2 node for qca9531 should be exactly the same as ar9344 and the cpuintc for qca9533.dtsi is missing ddr wb channels.
BTW There should be a qca9531/qca9533 split like ar9341/ar9344 because PCIE only exist on qca9531 and we can use cpuintc for wmac on qca9533.

juppin commented:
> BTW There should be a qca9531/qca9533 split like ar9341/ar9344 because PCIE only exist on qca9531 and we can use cpuintc for wmac on qca9533.

This mean the intc2 node should not exist on qca9533!
The kernel panic happens at the interrrupt driver init because the status reg addr isn´t set.

lucize commented:
> status reg addr isn´t set.

Yes I know, but with that set, it will crash with the first one, not panic

981213 commented:
It should be OK to add that.

lucize commented:
is like this

```
                        rst: reset-controller@1806001c {
                                compatible = "qca,qca9530-reset",
                                                "qca,ar7100-reset";
                                reg = <0x1806001c 0xac>;

                                #reset-cells = <1>;

                                intc2: interrupt-controller@2 {
                                        compatible = "qca,ar9340-intc";

                                        interrupt-parent = <&cpuintc>;
                                        interrupts = <2>;

                                        interrupt-controller;
                                        #interrupt-cells = <1>;

                                        qca,int-status-addr = <0xac>;
                                        qcom,pending-bits = <0x1f0>,    /* pcie rc1 */
                                                            <0xf>;      /* wmac */
                                };
                        };
```
```
                        wmac: wmac@18100000 {
                                compatible = "qca,qca9530-wmac";
                                reg = <0x18100000 0x230000>;

                                interrupt-parent = <&intc2>;
                                interrupts = <1>;

                                status = "disabled";
                        };
```

juppin commented:
> qcom,pending-bits = <0x1f0>,    /* pcie rc1 */
                                                            <0xf>;      /* wmac */

As i said, all properties are now qca insteaad of qcom!

981213 commented:
I do have a QCA9533 nearby. I'll try it now.
@lucize you should copy the ddr-wb-channels from ar9344, too.

lucize commented:
let me come back, I think that this is a 1/2 edited, and I tried qca also

juppin commented:
> let me come back, I think that this is a 1/2 edited, and I tried qca also

As @981213 simply copy the complete intc2 from ar9344.dtsi and use intc2 0 for wmac and intc2 1 for pci

lucize commented:
editing it like ar9344.dtsi seems to work ok

981213 commented:
@juppin I've got the ethernet problem solved. The resets for gmac0/gmac1 in ar934x is totally wrong.
```
&mdio0 {
	resets = <&rst 22>;
	reset-names = "mdio";
};

&eth0 {
	compatible = "qca,ar9340-eth", "syscon";

	pll-data = <0x16000000 0x00000101 0x00001616>;
	pll-reg = <0x4 0x2c 17>;
	pll-handle = <&pll>;

	resets = <&rst 8>, <&rst 9>;
	reset-names = "mac", "phy";
};

&mdio1 {
	resets = <&rst 23>;
	reset-names = "mdio";
	builtin-switch;
};

&eth1 {
	compatible = "qca,ar9340-eth", "syscon";

	resets = <&rst 12>, <&rst 13>;
	reset-names = "mac", "phy";
};
```
There is no phy reset for ar934x. rst9 is mac reset for gmac0 and rst 13 is mac reset for gmac1.
rst8 and rst12 are reset for the builtin switch.
Here is why I got everything broken even if I initialized gmac0 after gmac1: When initializing gmac0 we asserted the defined phy reset which in fact caused a reset to the builtin switch, and what we've configured for the builtin switch during initialization for gmac1 are all gone.

I'll create a large pull request for this and other things I mentioned previously later...

juppin commented:
@981213 You are right. I´ve seen this too that  mac and phy is swapped and there is no phy reset.
I´ve changed this in my local  wr841v8 (ar9341) banch some time ago.
I would commit this with my wr841v8 port, but this will take a while as i don´t figured out how to setup gmac0 properly.
Please create a pr of your fix.
